cpp/src/arrow/array.h

```diff
@@ -371,6 +370,33 @@ class ARROW_EXPORT FixedSizeBinaryArray : public PrimitiveArray {
 };
 
 // ----------------------------------------------------------------------
+// DecimalArray
+class ARROW_EXPORT DecimalArray : public FixedSizeBinaryArray {
+ public:
+  using TypeClass = Type;
+  DecimalArray(const std::shared_ptr<DataType>& type, int64_t length,
+      const std::shared_ptr<Buffer>& data,
+      const std::shared_ptr<Buffer>& null_bitmap = nullptr, int64_t null_count = 0,
+      int64_t offset = 0, const std::shared_ptr<PoolBuffer>& sign_bitmap = nullptr)
+      : FixedSizeBinaryArray(type, length, data, null_bitmap, null_count, offset),
+        sign_bitmap_(sign_bitmap),
+        sign_bitmap_data_(sign_bitmap != nullptr ? sign_bitmap->data() : nullptr) {}
+
+  bool IsNegative(int64_t i) const {
+    return sign_bitmap_data_ != nullptr ? BitUtil::GetBit(sign_bitmap_data_, i) : false;
+  }
+
+  template <typename T>
+  Decimal<T> Value(int64_t i) const;
+  std::shared_ptr<Array> Slice(int64_t offset, int64_t length) const override;
+
+ private:
+  /// Only needed for 128 bit Decimals
+  std::shared_ptr<PoolBuffer> sign_bitmap_ = nullptr;
```

Comment:
fixed, now uses `std::shared_ptr<Buffer>` instead of `std::shared_ptr<PoolBuffer>`.